test/unit/esm-named-exports.test.js

```diff
@@ -9,104 +9,122 @@ const expect = chai.expect;
-
-    const missingEsmKeys = [];
-    for (const key of cjsKeys) {
-      if (ignoredCjsKeys.includes(key)) {
-        continue;
+const exportPaths = ['@sequelize/core', '@sequelize/core/decorators-legacy'];
+
+// require('@sequelize/core') returns the Sequelize class
+// The typings do not reflect this as some properties of the Sequelize class are not declared as exported in types/index.d.ts.
+// This array lists the properties that are present on the class, but should not be exported in the esm export file nor in types/index.d.ts.
+const ignoredCjsKeysMap = {
+  '@sequelize/core': [
+    // make no sense to export
+    'length',
+    'prototype',
+    'name',
+    'version',
+
+    // importing the data type directly has been removed, and accessing them on the Sequelize constructor is deprecated.
+    // Use DataTypes.x exclusively.
+    'ABSTRACT',
+    'ARRAY',
+    'BIGINT',
+    'BLOB',
+    'BOOLEAN',
+    'CHAR',
+    'CIDR',
+    'CITEXT',
+    'DATE',
+    'DATEONLY',
+    'DECIMAL',
+    'DOUBLE',
+    'ENUM',
+    'FLOAT',
+    'GEOGRAPHY',
+    'GEOMETRY',
+    'HSTORE',
+    'INET',
+    'INTEGER',
+    'JSON',
+    'JSONB',
+    'MACADDR',
+    'MEDIUMINT',
+    'NOW',
+    'RANGE',
+    'REAL',
+    'SMALLINT',
+    'STRING',
+    'TEXT',
+    'TIME',
+    'TINYINT',
+    'TSVECTOR',
+    'UUID',
+    'UUIDV1',
+    'UUIDV4',
+    'VIRTUAL',
+  ],
+  '@sequelize/core/decorators-legacy': [
+    '__esModule',
+  ],
+};
+
+for (const exportPath of exportPaths) {
+  describe(`module "${exportPath}"`, () => {
+    it('exposes the same named exports as the CJS module', async () => {
+      // important: if you transpile this file, it's important
+      //  that we still use both the native import() and the native require().
+      //  don't transpile this import() to a require().
+      const sequelizeEsm = await import(exportPath);
+      const sequelizeCjs = require(exportPath);
+
+      const esmKeys = Object.keys(sequelizeEsm);
+
+      // include non-enumerables as "Sequelize.{and, or, ...}" are non-enumerable
+      const cjsKeys = Object.getOwnPropertyNames(sequelizeCjs);
+
+      const ignoredCjsKeys = ignoredCjsKeysMap[exportPath];
+      for (const key of ignoredCjsKeys) {
+        expect(cjsKeys)
+          .to
+          .include(key, `Sequelize static property ${JSON.stringify(key)} is marked as ignored for ESM export but does not exist. Remove it from ignore list.`);
       }
 
-      if (!esmKeys.includes(key)) {
-        missingEsmKeys.push(key);
-      }
-    }
+      const missingEsmKeys = [];
+      for (const key of cjsKeys) {
+        if (ignoredCjsKeys.includes(key)) {
+          continue;
+        }
 
-    expect(missingEsmKeys.length).to.eq(0,
-      `esm entry point is missing exports: ${missingEsmKeys.map(v => JSON.stringify(v)).join(', ')}.
-Either add these exports to "index.mjs" (and "types/index.d.ts"), or mark them as ignored in "esm-named-exports.test.js"
-      `);
+        if (!esmKeys.includes(key)) {
```

Comment:
We're looking for missing ESM keys here - do we want to check for any missing CJS ones for backwards compatibility (e.g. mistaken `export default` not being caught in code reviews)? It's not the express purpose of this file, but may be helpful. Happy to disregard if there's not a good reason for it though

Reply:
ESM can be checked automatically by comparing itself to CJS, but checking CJS would require manually listing all CJS exports in this file. Could be considered but not part of this PR

Reply:
Gotcha, resolving